### ISSUE TITLE
[no-unused-vars] False positive on generic inside merging declarations

### ISSUE THREAD

**Issue by saitonakamura** [typescript-eslint/typescript-eslint#3353]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": [
      "error",
      {
        "argsIgnorePattern": "^_",
      },
    ],
  }
}
```

`react.d.ts`

```TS
declare namespace React {
  interface IframeHTMLAttributes<T> {
    importance?: 'auto' | 'high' | 'low';
  }
}
```

**Expected Result**

`T` won't get reported on line 2 cause it's needed in this ambient context

**Actual Result**

Eslint reports on line 2 `'T' is declared but its value is never read`, but if you remove or rename it, TS gives you `All declarations of 'IframeHTMLAttributes' must have identical type parameters.`

**Additional Info**

The same things happens when using ambient contexts inside a `.ts` file, e.g.

test.ts

```ts
declare global {
  namespace jest {
    interface Matchers<R, T> {
      toBeCalledBefore: (expected: Mock<any, any>) => CustomMatcherResult;
      toBeCalledAfter: (expected: Mock<any, any>) => CustomMatcherResult;
    }
  }
}
```

`R` and `T` will be reported as `'R' is defined but never used`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.0` |
| `@typescript-eslint/parser`        | `4.220` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.25.0` |
| `node`                             | `14.16.0` |

**Comment by bradzacher** [comment]
There is no way for the lint rule to understand these sorts of typescript dependencies that occur across module boundaries.

You can use `varsIgnorePattern` to configure a naming convention to ignore variables, or you can use a disable comment

```ts
declare namespace React {
  interface IframeHTMLAttributes<_T> {
    importance?: 'auto' | 'high' | 'low';
  }
}

declare global {
  namespace jest {
    interface Matchers<_R, _T> {
      toBeCalledBefore: (expected: Mock<any, any>) => CustomMatcherResult;
      toBeCalledAfter: (expected: Mock<any, any>) => CustomMatcherResult;
    }
  }
}
```

**Comment by saitonakamura** [comment]
Thanks for the answer! I understand that you don't have this info in eslint, but maybe there is some heuristic we can apply to handle such cases, something like "generic variables in ambient contexts"? Or maybe we can disable this rule in ambient contexts by options, what do you think?

I also wanted to mention that you can't really use vars ignore pattern, because type parameters should be identical even by name across ambient contexts

**Comment by bradzacher** [comment]
> Or maybe we can disable this rule in ambient contexts by an option, what do you think?

You can do that already without any changes to the rule:
https://eslint.org/docs/user-guide/configuring/configuration-files#configuration-based-on-glob-patterns

**Comment by saitonakamura** [comment]
I was talking not only about `.d.ts` files, but `.ts` too. In `.ts` file you define an ambient context using `declare global`

**Comment by bradzacher** [comment]
in that context it's not safe to just ignore as you're more likely to have false negatives.
False negatives are a thousand times worse than false positives.

False positives can be seen, assessed and ignored.
False negatives are invisible and insidious. When found they erode a user's trust in linting because they don't know if the linter has missed other things.

It's a *very* rare case that a user will use an ambient module augmentation in a `.ts` file, and even rarer still that they will be augmenting an interface with generics that are not used in the outcome.

These are all the reasons that we will *not* be adding any handling for this case.

**Comment by saitonakamura** [comment]
Thank you for the explanation, it's really insightful!
